### PR TITLE
 [Cudo] Add new cloud Cudo compute using Provisioner method

### PR DESCRIPTION
**Pull request by JungleCatSW**
Refactored to use the Cudo Compute PR into a v2 provisioner.
There is also a fetch method as our catalog changes reguarly so there will be a pr on the catalog repo too.

Update:

Fixed memory filtering, copied external for multi node launches, fixed disk id issue, added a wait for correct state for terminate.

Tested (run the relevant ones):

- [x ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/backend_utils.py
Comment (Michaelvll) [inline sky/backends/backend_utils.py:1166]: nit:
```suggestion
    if isinstance(cloud, (clouds.AWS, clouds.OCI, clouds.SCP, cloud, clouds.Cudo)):
        config = auth.configure_ssh_info(config)
```

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:61]: Since we have this version set to `SKYPILOT`, we need to move the instance status query to the provisioner API, i.e. `sky/provision/cudo/instance.py`

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:106]: This is not required, as `instance_type is not None` is guaranteed by the caller.

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:95]: nit:
```suggestion
    def regions_with_offering(cls, instance_type: str,
```

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:92]: This is actually not needed.
```suggestion
```

File: sky/clouds/service_catalog/data_fetchers/requirements.txt
Comment (Michaelvll) [inline sky/clouds/service_catalog/data_fetchers/requirements.txt:7]: We probably don't want to have the cloud specific requirements listed here. We can have it in the `setup.py` for the extra `cudo`. : )

File: sky/provision/cudo/__init__.py
Comment (Michaelvll) [inline sky/provision/cudo/__init__.py:3]: This is not needed?
```suggestion
```

File: sky/provision/cudo/config.py
Comment (Michaelvll) [inline sky/provision/cudo/config.py:15]: This should not be needed.
```suggestion
```

File: sky/provision/cudo/__init__.py
Comment (Michaelvll) [inline sky/provision/cudo/__init__.py:4]: This is not needed?
```suggestion
```

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:354]: Seems this is already implemented in `sky/provision/cudo/instance.py`, it should be fine to be removed.
```suggestion
```

File: sky/setup_files/MANIFEST.in
Comment (Michaelvll) [inline sky/setup_files/MANIFEST.in:9]: This seems not needed, as we don't have node_provider for `cudo` now.

File: sky/templates/cudo-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/cudo-ray.yml.j2:77]: All fields after this line is no longer needed, as our new provisioner will handle the setup. Please refer to:
https://github.com/skypilot-org/skypilot/blob/9743aa0c7f86a352715b13dd18ee780239f190f6/sky/templates/aws-ray.yml.j2#L172-L174

**Review by Michaelvll — COMMENTED** [review]
Thank you for submitting the PR for support Cudo compute @JungleCatSW! The code looks fantastic and very clean. Just left several comments. : )

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:156]: nit: this should be fine to remove, considering we have the `self._REPR` set already, and the `Cloud.__repr__` should be able to handle this.

File: sky/clouds/service_catalog/data_fetchers/requirements.txt
Comment (Michaelvll) [inline sky/clouds/service_catalog/data_fetchers/requirements.txt:7]:
```suggestion
```

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:258]: nit:
```suggestion
                f'{self._INDENT_PREFIX}{common_utils.format_exception(e, use_bracket=True)}')
```

File: sky/clouds/cudo.py
Comment (Michaelvll) [inline sky/clouds/cudo.py:265]: nit:
```suggestion
                f'{self._INDENT_PREFIX}{common_utils.format_exception(e, use_bracket=True)}')
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick update @JungleCatSW! Left some minor comments. I think it should be good to go once we have done some basic tests after we get the access to the cloud.

File: sky/provision/cudo/instance.py
Comment (Michaelvll) [inline sky/provision/cudo/instance.py:158]:
```suggestion
        logger.debug(f'Removing {inst_id}: {inst}')
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick update @JungleCatSW! I am trying out the PR with the Cudo compute. Left several compute for the fix of the memory filtering.

I also encountered the following issue:
`sky launch -c test-cudo --gpus RTXA4000 --memory 10+ --num-nodes 4 echo hi`
The cluster provisioning fails due to an error related to the disk id:
```
D 01-16 19:03:30 provisioner.py:170]     provision_record = provision.run_instances(provider_name,
D 01-16 19:03:30 provisioner.py:170]   File "/home/gcpuser/skypilot/sky/provision/__init__.py", line 41, in _wrapper
D 01-16 19:03:30 provisioner.py:170]     return impl(*args, **kwargs)
D 01-16 19:03:30 provisioner.py:170]   File "/home/gcpuser/skypilot/sky/provision/cudo/instance.py", line 87, in run_instances
D 01-16 19:03:30 provisioner.py:170]     instance_id = cudo_wrapper.launch(
D 01-16 19:03:30 provisioner.py:170]   File "/home/gcpuser/skypilot/sky/provision/cudo/cudo_wrapper.py", line 44, in launch
D 01-16 19:03:30 provisioner.py:170]     raise e
D 01-16 19:03:30 provisioner.py:170]   File "/home/gcpuser/skypilot/sky/provision/cudo/cudo_wrapper.py", line 41, in launch
D 01-16 19:03:30 provisioner.py:170]     vm = api.create_vm(cudo().cudo_api.project_id(), request)
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/api/virtual_machines_api.py", line 487, in create_vm
D 01-16 19:03:30 provisioner.py:170]     (data) = self.create_vm_with_http_info(project_id, create_vm_body, **kwargs)  # noqa: E501
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/api/virtual_machines_api.py", line 557, in create_vm_with_http_info
D 01-16 19:03:30 provisioner.py:170]     return self.api_client.call_api(
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/api_client.py", line 326, in call_api
D 01-16 19:03:30 provisioner.py:170]     return self.__call_api(resource_path, method,
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/api_client.py", line 158, in __call_api
D 01-16 19:03:30 provisioner.py:170]     response_data = self.request(
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/api_client.py", line 368, in request
D 01-16 19:03:30 provisioner.py:170]     return self.rest_client.POST(url,
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/rest.py", line 269, in POST
D 01-16 19:03:30 provisioner.py:170]     return self.request("POST", url,
D 01-16 19:03:30 provisioner.py:170]   File "/opt/conda/envs/sky/lib/python3.10/site-packages/cudo_compute/rest.py", line 228, in request
D 01-16 19:03:30 provisioner.py:170]     raise ApiException(http_resp=r)
D 01-16 19:03:30 provisioner.py:170] cudo_compute.rest.ApiException: (409)
D 01-16 19:03:30 provisioner.py:170] Reason: Conflict
D 01-16 19:03:30 provisioner.py:170] HTTP response headers: HTTPHeaderDict({'Date': 'Tue, 16 Jan 2024 19:03:30 GMT', 'Content-Type': 'application/json', 'Content-Length': '70', 'Connection': 'keep-alive', 'vary': 'Origin', 'CF-Cache-Status': 'DYNAMIC', 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=5iQ83c%2B9r7u6IVUB0XNJ8w6S%2FwN0Wyt%2F7G%2FSduK%2Fq16YJvwePFNSg%2FHD%2BooFs4uJkl9BxGSPH19jdHOyOfJbknvgFoXEx8diTVqSW2e4vO148p1QL%2BZYy7XraojND0QqT7YJID%2FozGk%3D"}],"group":"cf-nel","max_age":604800}', 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload', 'X-Content-Type-Options': 'nosniff', 'Server': 'cloudflare', 'CF-RAY': '84689e682d8e5dd6-HKG', 'alt-svc': 'h3=":443"; ma=86400'})
D 01-16 19:03:30 provisioner.py:170] HTTP response body: {"code":6,"message":"A disk with that id already exists","details":[]}
```
This should be fine if it is a transient issue with the cloud, but a more serious problem is the resource leakage caused by the failure of the auto termination triggered by SkyPilot. After the error is raised above, SkyPilot will call the `sky.provision.cudo.instance::terminate_instance` to terminate the partially ready VMs, and the following error occurs:
```
D 01-16 19:03:30 provisioner.py:179] Terminating the failed cluster.
D 01-16 19:03:31 cloud_vm_ray_backend.py:1152] Got error(s) in Cudo:[cudo_compute.rest.ApiException] (400)
D 01-16 19:03:31 cloud_vm_ray_backend.py:1152] Reason: Bad Request
D 01-16 19:03:31 cloud_vm_ray_backend.py:1152] HTTP response headers: HTTPHeaderDict({'Date': 'Tue, 16 Jan 2024 19:03:31 GMT', 'Content-Type': 'application/json', 'Content-Length': '211', 'Connection': 'keep-alive', 'vary': 'Origin', 'CF-Cache-Status': 'DYNAMIC', 'Report-To': '{"endpoints":[{"url":"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=PuvVoBDN1Qo8X3NHl0oqIc02Ldsh%2B1c2OeOgafxQhBCC8vg7fJlUJHoM%2Bh1DrdZHWgSsYnQWV4YrgkFal0FFih7PhyYr5TokcIpc1qjl6bwTSW51jBKF%2FXHnzkGJ4psfANopLiyGnUU%3D"}],"group":"cf-nel","max_age":604800}', 'NEL': '{"success_fraction":0,"report_to":"cf-nel","max_age":604800}', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload', 'X-Content-Type-Options': 'nosniff', 'Server': 'cloudflare', 'CF-RAY': '84689e6facad04ce-HKG', 'alt-svc': 'h3=":443"; ma=86400'})
D 01-16 19:03:31 cloud_vm_ray_backend.py:1152] HTTP response body: {"code":9,"message":"Invalid vm state (prol)","details":[{"@type":"type.googleapis.com/google.rpc.PreconditionFailure","violations":[{"type":"","subject":"","description":"vm cannot currently be terminated"}]}]}
```

It seems to be a problem for not being able to call termination on a VM with state `prol`. Should we wait until the instance to get out of the pending stage before we call  termination operation?

File: sky/provision/cudo/instance.py
Comment (Michaelvll) [inline sky/provision/cudo/instance.py:219]:
```suggestion
    cluster_name_on_cloud: str,
    ports: List[str],
    provider_config: Optional[Dict[str, Any]] = None,
```

**Comment by Michaelvll** [comment]
Another issue: it seems when I try to launch a multi-node cluster with `sky launch --num-nodes 2 --cloud cudo --gpus RTXA4000`, SkyPilot gets empty internal IPs for the VM. This causes the failure for starting the ray cluster using the internal IPs. Here is an InstanceInfo for a VM printed by the provisioner `{'status': 'runn', 'tags': {}, 'name': 'test-cudo-2514-head', 'ip': '149.36.1.9', 'external_ip': '149.36.1.9', 'internal_ip': ''}` 

Do we have an idea why the internal IP is empty and how can we fix that? If it is fine, we can also set the internal IP to be the same as external IP, connecting the VMs directly using the extneral IPs.

File: sky/clouds/cudo.py
Comment (JungleCatSW) [reply sky/clouds/cudo.py:224]: Would it be okay to poll and wait within the terminate instance function ? Also I will work on the other comments thanks!

**Comment by JungleCatSW** [comment]
Everything except the disk id issue is resolved, I need to make some changes elsewhere to fix that. I will let you know when it is ready.

File: sky/provision/cudo/instance.py
Comment (Michaelvll) [inline sky/provision/cudo/instance.py:109]: This should be `created_instance_ids` instead?
```suggestion
    for instance_id in created_instance_ids:
```

File: sky/provision/cudo/instance.py
Comment (Michaelvll) [inline sky/provision/cudo/instance.py:107]: I tried `sky launch --num-nodes 2 --gpus RTXA4000 --cloud cudo` and it seems the VM is still in `starting` state after more than 6 minutes. Probably the 120 seconds wait time is not enough? Similar thing happens to RTX3080

Comment (Michaelvll) [reply sky/provision/cudo/instance.py:107]: Update: it seems to be a cloud backend issue as `sky launch --gpus RTXA4000 --cloud cudo` also took significant amount of time, but `RTXA6000` seems fine.

**Review by Michaelvll — APPROVED** [review]
Thank you for the quick fix @JungleCatSW! This is fantastic! I just tried launching the single and multiple VMs and it seems working correctly. We are happy to merge this PR in first with the comments below fixed.

After the PR is merged, there will be several TODOs:
1. Add instruction for how to setup the cudo credentials in our https://github.com/skypilot-org/skypilot/blob/master/docs/source/getting-started/installation.rst
2. Add tests in our tests: https://github.com/skypilot-org/skypilot/blob/master/tests/conftest.py#L23-L26

File: sky/provision/cudo/instance.py
Comment (JungleCatSW) [reply sky/provision/cudo/instance.py]: I have adjusted it to 10 minutes, although they should usually launch in under a minute. I will have to do some investigation there. Let me know if 10 minutes is too long...

**Comment by Michaelvll** [comment]
Could we run the `format.sh` to format the code?

**Pull request merged**
